src/lightning/fabric/plugins/precision/bitsandbytes.py

```diff
@@ -0,0 +1,264 @@
+
+        mode_to_cls = {
+            "nf4": _NF4Linear,
+            "nf4-dq": _NF4DQLinear,
+            "fp4": _FP4Linear,
+            "fp4-dq": _FP4DQLinear,
+            "int8-training": _Linear8bitLt,
+            "int8": _Int8LinearInference,
+        }
+        self._linear_cls = mode_to_cls[mode]
+        self.dtype = dtype
+        self.skips = skips or set()
+
+    def convert_module(self, module: torch.nn.Module) -> torch.nn.Module:
+        # avoid naive users thinking they quantized their model
+        if not any(isinstance(m, torch.nn.Linear) for m in module.modules()):
+            raise TypeError(
+                "You are using the bitsandbytes precision plugin, but your model has no Linear layers. This plugin"
+                " won't work for your model."
+            )
+
+        # convert modules if they haven't been converted already
+        if not any(
+            "bitsandbytes.nn" in m.__module__ or "precision.bitsandbytes" in m.__module__ for m in module.modules()
+        ):
+            _convert_layers(module, self._linear_cls, self.skips)
+
+        # set the compute dtype if necessary
+        for m in module.modules():
+            if isinstance(m, _Linear4bit):
+                m.compute_dtype = self.dtype
+                m.compute_type_is_set = False
+        return module
+
+    def init_context(self) -> ContextManager:
+        dtype_ctx = _DtypeContextManager(self.dtype)
+        if self.skips:
+            # cannot patch the Linear class if the user wants to skip some submodules
+            return dtype_ctx
+        stack = ExitStack()
+        stack.enter_context(dtype_ctx)
+        # TODO: this could also consider replacing with `bnb.nn.Embedding`
+        context_manager = _ClassReplacementContextManager({"torch.nn.Linear": self._linear_cls})
+        stack.enter_context(context_manager)
+        return stack
+
+    def forward_context(self) -> ContextManager:
+        return _DtypeContextManager(self.dtype)
+
+    def convert_input(self, data: Any) -> Any:
+        return apply_to_collection(data, function=_convert_fp_tensor, dtype=Tensor, dst_type=self.dtype)
+
+    def convert_output(self, data: Any) -> Any:
+        return apply_to_collection(data, function=_convert_fp_tensor, dtype=Tensor, dst_type=torch.get_default_dtype())
+
+
+def _import_bitsandbytes() -> ModuleType:
+    if not _BITSANDBYTES_AVAILABLE:
+        raise ModuleNotFoundError(str(_BITSANDBYTES_AVAILABLE))
+    # configuration for bitsandbytes before import
+    os.environ["BITSANDBYTES_NOWELCOME"] = "1"
+    warnings.filterwarnings("ignore", message=r".*bitsandbytes was compiled without GPU support.*")
+    warnings.filterwarnings(
+        "ignore", message=r"MatMul8bitLt: inputs will be cast from .* to float16 during quantization"
+    )
+    import bitsandbytes
+
+    return bitsandbytes
+
+
+if _BITSANDBYTES_AVAILABLE:
+    bnb = _import_bitsandbytes()
```

Comment:
If bitsandbytes is installed but not used by the user, this code would still run at import time of lightning. Could we avoid this? The idea is that

- the two utility functions below don't need to be guarded
- the layer definitions could be moved to a function and be returned as a dict, from which you then choose the one you need

Reply:
> the layer definitions could be moved to a function and be returned as a dict, from which you then choose the one you need

That is a strange pattern. How does this work under the hood? Does Python know that the class definition is the same across calls? Or would each call create new instances with the same name?

To be clear, you are suggesting?
```
def bnb_linears():
    import bitsandbytes as bnb

	class _Linear4bit(bnb.nn.Linear4bit):
	    ...

    ...

	return {"fp4": _Linear4bit, ...}
```

Reply:
No I guess it's not good to return new class objects everytime the function is called. All I'm saying is we should try to make the definition lazy. Maybe with `__getattr__` defined on the module. 

Not a blocker for this PR but we should try it, even in a follow up is fine with me. 
Because we've seen in the past how importing these 3rd party libraries ahead of time can have unintended side effects or slowdowns for users who don't even use these features. Examples from the past are deepspeed, torchelastic, logger integration libraries, torchtext etc.

Reply:
Let me try __getattr__